src/_arraykit.c

```diff
@@ -383,7 +383,28 @@ dtype_from_element(PyObject *Py_UNUSED(m), PyObject *arg)
 
     // Integers
     if (PyLong_CheckExact(arg)) {
-        return (PyObject*)PyArray_DescrFromType(NPY_LONG);
+        int overflow;
+        npy_int64 v = PyLong_AsLongLongAndOverflow(arg, &overflow);
+        if (v != -1 && overflow == 0) {
+            // Prioritize fastest case first
+            return (PyObject*)PyArray_DescrFromType(NPY_LONG); // (-2**63, 2**63)
+        }
+        if (v == -1 && PyErr_Occurred()) {
+            return NULL;
+        }
+        if (overflow == -1) {
+            return (PyObject*)PyArray_DescrFromType(NPY_OBJECT); // [-∞, -2**63]
+        }
+
+        npy_uint64 uv = PyLong_AsUnsignedLongLong(arg);
+        if (uv == (npy_uint64)-1 && PyErr_Occurred()) {
```

Comment:
I was staring at that cast for *way* too long thinking it was some sort of subtraction. I prefer this form:

```suggestion
        if (uv == -1ULL && PyErr_Occurred()) {
```